azure/coverage_job.yml

```diff
@@ -0,0 +1,67 @@
+parameters:
+  configurationName: ''
+  configurationParameters: ''
+  runTestsParameters: ''
+  timeoutInMinutes: 60
+
+jobs:
+  - job: ${{ parameters.configurationName }}
+    timeoutInMinutes: ${{ parameters.timeoutInMinutes }}
+    pool:
+      vmImage: 'ubuntu-latest'
+    steps:
+    - template: apt.yml
+      parameters:
+        packages: lcov
+    - script: |
+        sudo -H pip install gcovr
+    - template: configure.yml
+      parameters:
+        configurationParameters: ${{ parameters.configurationParameters }}
+    - script: make -j$(/usr/bin/nproc) >/dev/null
+      displayName: 'Make Build'
+    - script: |
+        sudo make install
+        sudo mkdir     /etc/php.d
+        sudo chmod 777 /etc/php.d
+        echo mysqli.default_socket=/var/run/mysqld/mysqld.sock     > /etc/php.d/mysqli.ini
+        echo pdo_mysql.default_socket=/var/run/mysqld/mysqld.sock  > /etc/php.d/pdo_mysql.ini
+        echo opcache.enable_cli=1 >> /etc/php.d/opcache.ini
+        echo opcache.protect_memory=1 >> /etc/php.d/opcache.ini
+      displayName: 'Install Build'
+    - script: |
+        mysql -uroot -proot -e "CREATE DATABASE IF NOT EXISTS test"
+        sudo -u postgres psql -c "ALTER USER postgres PASSWORD 'postgres';"
+        sudo -u postgres psql -c "CREATE DATABASE test;"
+      displayName: 'Setup'
+    - template: test.yml
+      parameters:
+        configurationName: ${{ parameters.configurationName }}
+        runTestsParameters: ${{ parameters.runTestsParameters }}
+    - template: test.yml
+      parameters:
+        configurationName: ${{ parameters.configurationName }}
+        runTestsName: 'OpCache'
+        runTestsParameters: >-
+          ${{ parameters.runTestsParameters }}
+          -d zend_extension=opcache.so
+    - script: |
+        rm -rf coverage.xml | true
+        gcovr -sr . -o coverage.xml --xml \
+          -e '.*/ext/bcmath/libbcmath/.*' \
+          -e '.*/ext/date/lib/.*' \
+          -e '.*/ext/fileinfo/libmagic/.*' \
+          -e '.*/ext/gd/libgd/.*' \
+          -e '.*/ext/hash/sha3/.*' \
+          -e '.*/ext/mbstring/libmbfl/.*' \
+          -e '.*/ext/opcache/jit/libudis86/.*' \
+          -e '.*/ext/pcre/pcre2lib/.*' \
+          -e '.*/ext/xmlrpc/libxmlrpc/.*'
```

Comment:
Based on https://dev.azure.com/phpazuredevops/PHP/_build/results?buildId=2971&view=codecoverage-tab, it doesn't look like these excludes are working. Docs are not super clear, but possibly the leading `.*/` is too much.

Reply:
Yes, gcovr is emitting warnings (see log at below). 

The date lib (`ext/date/lib`) causes warnings, in lcov and gcovr, related not being able to open file on `<stdout>`. I don't know reason for these for these warnings. In `lcov` these warnings can be prevented by excluding `*/<stdout>`:

```
$(LTP) --output-file $@ --remove $@ '*/<stdout>' $(LCOV_EXCLUDES)
```

In gcovr we need to use the `--exclude-directories`:

```gcovr -sr . -o gcovr.xml --xml \
    --exclude-directories 'ext/date/lib$$' \
    $(foreach lib, $(GCOVR_EXCLUDES), -e $(lib))
```

## Log

```
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/2d277420-04f0-44e9-95fb-913a6b2e7ce5.sh
(WARNING) GCOV produced the following errors processing /home/vsts/work/1/s/ext/date/lib/parse_iso_intervals.gcda:
    Cannot open source file parse_iso_intervals.re
Cannot open source file <stdout>

    Cannot open source file parse_iso_intervals.re
Cannot open source file <stdout>

    Cannot open source file <stdout>

    Cannot open source file parse_iso_intervals.re
Cannot open source file <stdout>

    Cannot open source file parse_iso_intervals.re
Cannot open source file <stdout>

    Cannot open source file parse_iso_intervals.re
Cannot open source file <stdout>

    Cannot open source file parse_iso_intervals.re
Cannot open source file <stdout>

    Cannot open source file parse_iso_intervals.re
Cannot open source file <stdout>

    Cannot open source file parse_iso_intervals.re
Cannot open source file <stdout>

    Could not open output file 'parse_iso_intervals.re.gcov'
Could not open output file '<stdout>.gcov'

    Cannot open source file parse_iso_intervals.re
Cannot open source file <stdout>

    (gcovr could not infer a working directory that resolved it.)
(WARNING) GCOV produced the following errors processing /home/vsts/work/1/s/ext/date/lib/parse_date.gcda:
    Cannot open source file parse_date.re
Cannot open source file <stdout>

    Cannot open source file parse_date.re
Cannot open source file <stdout>

    Cannot open source file <stdout>

    Cannot open source file parse_date.re
Cannot open source file <stdout>

    Cannot open source file parse_date.re
Cannot open source file <stdout>

    Cannot open source file parse_date.re
Cannot open source file <stdout>

    Cannot open source file parse_date.re
Cannot open source file <stdout>

    Cannot open source file parse_date.re
Cannot open source file <stdout>

    Cannot open source file parse_date.re
Cannot open source file <stdout>

    Could not open output file 'parse_date.re.gcov'
Could not open output file '<stdout>.gcov'

    Cannot open source file parse_date.re
Cannot open source file <stdout>

    (gcovr could not infer a working directory that resolved it.)
lines: 72.1% (189432 out of 262876)
branches: 51.6% (123168 out of 238880)
```